OpenSim/Common/XsensDataReader.cpp

```diff
@@ -0,0 +1,224 @@
+            it != imuStreams.end();
+            ++it, ++imu_index) {
+            std::ifstream* nextStream = *it;
+            // parse gyro info from imuStream
+            std::vector<std::string> nextRow = FileAdapter::getNextLine(*nextStream, "\t\r");
+            if (nextRow.empty()) {
+                done = true;
+                break;
+            }
+            if (foundLinearAccelerationData)
+                accel_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[accIndex]),
+                    std::stod(nextRow[accIndex + 1]), std::stod(nextRow[accIndex + 2]));
+            if (foundMagneticHeadingData)
+                magneto_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[magIndex]),
+                    std::stod(nextRow[magIndex + 1]), std::stod(nextRow[magIndex + 2]));
+            if (foundAngularVelocityData)
+                gyro_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[gyroIndex]),
+                    std::stod(nextRow[gyroIndex + 1]), std::stod(nextRow[gyroIndex + 2]));
+            // Create Mat33 then convert into Quaternion
+            SimTK::Mat33 imu_matrix{ SimTK::NaN };
+            int matrix_entry_index = 0;
+            for (int mcol = 0; mcol < 3; mcol++) {
+                for (int mrow = 0; mrow < 3; mrow++) {
+                    imu_matrix[mrow][mcol] = std::stod(nextRow[rotationsIndex + matrix_entry_index]);
+                    matrix_entry_index++;
+                }
+            }
+            // Convert imu_matrix to Quaternion
+            SimTK::Rotation imu_rotation{ imu_matrix };
+            orientation_row_vector[imu_index] = imu_rotation.convertRotationToQuaternion();
+        }
+        if (done) 
+            break;
+        // append to the tables
+        times[rowNumber] = time;
+        if (foundLinearAccelerationData) 
+            linearAccelerationData[rowNumber] =  accel_row_vector;
+        if (foundMagneticHeadingData) 
+            magneticHeadingData[rowNumber] = magneto_row_vector;
+        if (foundAngularVelocityData) 
+            angularVelocityData[rowNumber] = gyro_row_vector;
+        rotationsData[rowNumber] = orientation_row_vector;
+        time += timeIncrement;
+        rowNumber++;
+        if (std::remainder(rowNumber, last_size) == 0) {
+            // resize all Data/Matrices, double the size  while keeping data
+            int newSize = last_size*2;
+            times.resize(newSize);
+            // Repeat for Data matrices in use
+            if (foundLinearAccelerationData) linearAccelerationData.resizeKeep(newSize, n_imus);
+            if (foundMagneticHeadingData) magneticHeadingData.resizeKeep(newSize, n_imus);
+            if (foundAngularVelocityData) angularVelocityData.resizeKeep(newSize, n_imus);
+            rotationsData.resizeKeep(newSize, n_imus);
+            last_size = newSize;
+        }
+    }
+    // Trim Matrices in use to actual data and move into tables
+    int actualSize = rowNumber;
```

Comment:
Why introduce `actualSize`? Simpler to just use `rowNumber` or call `rowCount`.

Reply:
@aseth1 Agreed it could be done this way, but I found it useful in debugging to have a separate variable for actualSize to avoid off-by-one mistakes, especially when looking at loops when you don't know if count starts at 0 or 1 due to bootstrapping or when count is incremented. If you find it confusing I can still do it.

Reply:
Ok, I removed the variable altogether and used rowNumber